File: apps/mobile/lib/src/features/vocabulary/domain/review_scheduler.dart
import 'review_result.dart';

class ReviewScheduler {
  const ReviewScheduler();

  ReviewResult schedule({
    required bool known,
    required int currentReviewCount,
    required DateTime reviewedAt,
  }) {
    if (currentReviewCount < 0) {
      throw ArgumentError('currentReviewCount must be zero or positive');
    }

    final nextCount = currentReviewCount + 1;
    final intervalDays = known ? _knownIntervalDays(nextCount) : 1;
    return ReviewResult(
      reviewStatus: known && nextCount >= 4 ? 'known' : 'learning',
      reviewCount: nextCount,
      nextReviewAt: reviewedAt.toUtc().add(Duration(days: intervalDays)),
      lastReviewedAt: reviewedAt.toUtc(),
    );
  }

  int _knownIntervalDays(int nextReviewCount) {
    return switch (nextReviewCount) {
      1 => 3,
      2 => 7,
      3 => 15,
      _ => 30,
    };
  }
}

File: apps/web-reader/services/reviewScheduler.ts
import type { WebReviewStatus } from '../types/localData'

export interface ReviewScheduleInput {
  known: boolean
  reviewCount: number
  reviewedAt: string
}

export interface ReviewScheduleResult {
  reviewStatus: WebReviewStatus
  reviewCount: number
  reviewedAt: string
  lastReviewedAt: string
  nextReviewAt: string
}

export function scheduleReview(input: ReviewScheduleInput): ReviewScheduleResult {
  const nextReviewCount = input.reviewCount + 1
  const reviewedAt = new Date(input.reviewedAt)
  const intervalDays = input.known ? knownIntervalDays(input.reviewCount) : 1
  const nextReviewAt = new Date(reviewedAt)
  nextReviewAt.setUTCDate(nextReviewAt.getUTCDate() + intervalDays)

  return {
    reviewStatus: input.known ? statusForKnownReview(nextReviewCount) : 'learning',
    reviewCount: nextReviewCount,
    reviewedAt: input.reviewedAt,
    lastReviewedAt: input.reviewedAt,
    nextReviewAt: nextReviewAt.toISOString(),
  }
}

function knownIntervalDays(previousReviewCount: number): number {
  if (previousReviewCount <= 0) {
    return 3
  }
  if (previousReviewCount === 1) {
    return 7
  }
  if (previousReviewCount === 2) {
    return 15
  }
  return 30
}

function statusForKnownReview(reviewCount: number): WebReviewStatus {
  return reviewCount >= 4 ? 'mastered' : 'reviewing'
}
